Added basic weather functionality

jasonw4331 opened the pull request:
- [x] Add weather commands
- [x] Add lightning strikes for thunderstorms with proper calculations
- [x] Add saving to level.dat via levelData in the provider
- [x] Synchronize existing level weather data to generating level
- [x] Send client weather on join
- [x] Implement rainLevel in StartGamePacket();
- [ ] Implement lightningLevel in StartGamePacket();
- [ ] Implement data in LevelEventPacket();

jasonw4331 committed:
Added basic weather functionality
Added weather commands
Added lightning for thunderstorms

jasonw4331 committed:
Corrected improper indentation

jasonw4331 committed:
Moved sendWeather() to Player.php
corrected the chance of lightning striking in a storm
corrected formatting to PM style

dktapps commented on src/pocketmine/Player.php:
```diff
@@ -1776,7 +1778,7 @@ protected function processLogin(){
 		$pk->dayCycleStopTime = -1; //TODO: implement this properly
 		$pk->eduMode = 0;
-		$pk->rainLevel = 0; //TODO: implement these properly
+		$pk->rainLevel = 0; //TODO: set based on precipitation of biome and height in world
```
Are you sure about this? I was under the impression these are global weather strengths. Modification based on biome and height is done on the client side.

jasonw4331 committed:
Merge remote-tracking branch 'pmmp/master' into weather

jasonw4331 committed:
Added level.dat saving
Corrected Commands
Removed Duration from WeatherChangeEvent
Changed spaced indents into tab indents

jasonw4331 committed:
forgot to remove the setWeather function

jasonwynn10 commented:
So... the different weather times count down until the weather begins. Also, rain can last from 1/2 a minecraft day to 7 1/2 minecraft days

jasonw4331 committed:
Merge remote-tracking branch 'pmmp/master' into weather

jasonw4331 committed:
forgot to remove this...

jasonwynn10 commented:
does whyone have any idea as to what the data value should be in the levelEventPacket?

jasonwynn10 commented:
Does anyone have any idea as to what the data value should be for weather in the levelEventPacket?

TheDiamondYT1 commented on src/pocketmine/command/defaults/ToggleDownfall.php:
```diff
@@ -0,0 +1,73 @@
+				}
+				Command::broadcastCommandMessage($sender, "Toggled downfall", true);
+				return true;
+				break;
```
Is this break needed?

jasonwynn10 replied:
no, but it's not hurting anything

TheDiamondYT1 commented on src/pocketmine/command/defaults/WeatherCommand.php:
```diff
@@ -0,0 +1,86 @@
+				$level->setClearTime($duration);
+				Command::broadcastCommandMessage($sender, "Changing to rain and thunder", true);
+				return true;
+				break;
```
Again, are these break statements needed?

TheDiamondYT1 commented on src/pocketmine/level/Level.php:
```diff
@@ -128,9 +130,12 @@ class Level implements ChunkManager, Metadatable{
-
 	const TIME_FULL = 24000;
 
+	const WEATHER_NORM = 0;
```
`WEATHER_NORMAL` would be a better name.

jasonwynn10 replied:
what about `WEATHER_CLEAR`?

thebigsmileXD replied:
Sky

TheDiamondYT1 replied:
@jasonwynn10 seems good

jasonwynn10 replied:
Changed in 6140c3a

TheDiamondYT1 commented on src/pocketmine/level/Level.php:
```diff
@@ -723,6 +745,40 @@ public function doTick(int $currentTick){
+			if($this->weather = self::WEATHER_RAIN_THUNDER){
+				foreach($this->getChunks() as $chunk){
+					if(mt_rand(1, 100000) == 1){
+						$pk = new AddEntityPacket();
```
It would be best to create an entity class for Lightning, since it *is* an entity. Either in this commit or when the entity rewrite comes @dktapps

thebigsmileXD replied:
Aren't lightnings also "doable" using a leveleventpacket?

TheDiamondYT1 replied:
In PC they are

TheDiamondYT1 replied:
No, but [you can start a thunder storm which probably spawns lightning with it](https://github.com/pmmp/PocketMine-MP/blob/api3/blocks/src/pocketmine/network/mcpe/protocol/LevelEventPacket.php#L77)

jasonwynn10 replied:
@thebigsmileXD how would you trigger lightning using a LevelEventPacket?

thebigsmileXD replied:
I know it was possible back in 0.15, i don't know if it still is a thing.

jasonwynn10 replied:
If you can figure it out, I'll use it

TheDiamondYT1 replied:
I can give you a 0.15 apk i have 25 different mcpe apks

dktapps replied:
Whether this was possible in 0.15 is irrelevant.

thebigsmileXD replied:
> I can give you a 0.15 apk i have 25 different mcpe apks

Advertising illegal App sharing? Idk.. and sure i have old versions too. i am a dev.

@dktapps seems like they removed the lightning level event, or changed it.

jasonwynn10 replied:
maybe I'll leave the entity for the rewrite...

thebigsmileXD replied:
Due to lightnings not despawning this will lead into an excessive amount of fake lightnings on the client side that reappear when reentering the chunk after its been unloaded client side, causing horrendous amounts of lags because most phones are not capable of rendering more then 20 lightnings.

jasonwynn10 replied:
@thebigsmileXD what would you have me do then?

thebigsmileXD replied:
I'll recheck and inform you if this can be done with a level event, if not.. adding to lightning.php that it dies after xxx ticks

TheDiamondYT1 commented on src/pocketmine/level/Level.php:
```diff
@@ -2713,6 +2772,107 @@ public function setSeed(int $seed){
+	 *
+	 * @return int
+	 */
+	public function getClearTime() {
```
Typehint?

TheDiamondYT1 commented on src/pocketmine/level/Level.php:
```diff
@@ -2713,6 +2772,107 @@ public function setSeed(int $seed){
+	}
+
+	/**
+     * Sets wether or not Weather is updated each tick
```
Formatting :)

jojoe77777 replied:
Also *whether

TheDiamondYT1 replied:
And possibly would be better to be `Sets whether or not the weather is updated each tick`. Notice the added `the`

TheDiamondYT1 commented on src/pocketmine/level/format/io/LevelProvider.php:
```diff
@@ -185,6 +185,18 @@ public function getTime();
 	public function setTime($value);
 
+    /**
+     * @return int The integer version of the type
```
Formatting

TheDiamondYT1 commented on src/pocketmine/Player.php:
```diff
@@ -3034,6 +3036,24 @@ public function sendTranslation($message, array $parameters = []){
+        $pk1->data = 90000; //TODO find correct values
+        $pk2->data = 90000;
+        $this->dataPacket($pk1);
+        $this->dataPacket($pk2);
```
If the weather is rain, you're sending a LevelEventPacket without an event id.

jasonwynn10 replied:
fixed in e94036d

jasonw4331 committed:
Merge branch 'master' into weather

# Conflicts:
#	src/pocketmine/Player.php
#	src/pocketmine/command/SimpleCommandMap.php
#	src/pocketmine/level/Level.php

jasonw4331 committed:
Fixed formatting and added type hints
Filled in evid field in LevelEventPacket 2 for rain

jasonwynn10 commented on src/pocketmine/Player.php:
```diff
@@ -3590,6 +3591,25 @@ public function sendTranslation($message, array $parameters = []){
+            $pk2->evid = LevelEventPacket::EVENT_START_THUNDER;
+        }elseif($this->level->getWeather() == Level::WEATHER_RAIN){
+            $pk1->evid = LevelEventPacket::EVENT_START_RAIN;
+            $pk2->evid = LevelEventPacket::EVENT_STOP_THUNDER;
```
will sending the stop thunder evid cause glitches if the thunder wasn't already thundering?

jasonw4331 committed:
Fixed more formatting
Corrected field name from AddEntityPacket
Fixed a glitch where weather wasn't sent when a player joined
Added fire to lightning

jasonw4331 committed:
Modified thunder duration
Added weather to the generation in McRegion
Synced weather between levels and their providers

jasonw4331 committed:
Fixed a mistake causing Levels to not generate

jasonw4331 committed:
I sincerely hate formatting...

jasonw4331 committed:
Changed Weather_NORM constant to WEATHER_CLEAR

jasonw4331 committed:
Utilized fields in StartGamePacket
Allow lightning to work while weather is locked

jasonw4331 committed:
Requested changes have been made. Merge?

sylvrs reviewed (approved):
Nice job!

xinghao2003 commented:
Good

jasonwynn10 commented:
@dktapps I believe this is good to merge if you are okay with it. 👍

TheDiamondYT1 commented on src/pocketmine/command/defaults/ToggleDownfall.php:
```diff
@@ -0,0 +1,78 @@
+use pocketmine\level\Level;
+use pocketmine\Player;
+
+class ToggleDownfall extends VanillaCommand{
```
To match **every other command:**

`ToggleDownfallCommand`

TheDiamondYT1 commented on src/pocketmine/level/Level.php:
```diff
@@ -782,6 +805,42 @@ public function doTick(int $currentTick){
+			--$this->thunderTime;
+			--$this->rainTime;
+			--$this->clearTime;
+			if($this->weather == self::WEATHER_CLEAR and ($this->rainTime <= 0 or $this->thunderTime <= 0)){
```
Sorry to be annoying...but you have a mixture of `==` and `===`. Do you mind changing them to the latter if possible?

Othee than that, LGTM

dktapps commented on src/pocketmine/Player.php:
```diff
@@ -3590,6 +3591,25 @@ public function sendTranslation($message, array $parameters = []){
 		$this->dataPacket($pk);
 	}
 
+	public function sendWeather(){
```
I don't like this here. It adds extra overhead for broadcasting purposes. Player also doesn't need any more mess in it.

TheDiamondYT1 replied:
Where else would it go?

dktapps replied:
This is core Level functionality, not Player. It should be broadcasted in the same way that time is.

TheDiamondYT1 replied:
Even Tess**** got this right. Wow.

dktapps commented on src/pocketmine/command/defaults/ToggleDownfall.php:
```diff
@@ -0,0 +1,78 @@
+				$level->setThunderTime($duration * 3);
+				Command::broadcastCommandMessage($sender, "Toggled downfall", true);
+				return true;
+				break;
```
This is unneeded.

dktapps commented on src/pocketmine/command/defaults/ToggleDownfall.php:
```diff
@@ -0,0 +1,78 @@
+				$level->setClearTime($duration);
+				Command::broadcastCommandMessage($sender, "Toggled downfall", true);
+				return true;
+				break;
```
also unnecessary

dktapps commented on src/pocketmine/command/defaults/ToggleDownfall.php:
```diff
@@ -0,0 +1,78 @@
+			default:
+				if(mt_rand(0, 100) > 94){
+					$level->setWeather(Level::WEATHER_RAIN_THUNDER);
+				} else {
```
formatting

dktapps commented on src/pocketmine/command/defaults/WeatherCommand.php:
```diff
@@ -0,0 +1,86 @@
+				$level->setClearTime($duration);
+				Command::broadcastCommandMessage($sender, "Changing to rain and thunder", true);
+				return true;
+				break;
```
these useless breaks are everywhere... why?

dktapps commented on src/pocketmine/entity/Entity.php:
```diff
@@ -1166,7 +1166,7 @@ public function entityBaseTick($tickDiff = 1){
 			}
 
-			if($this->fireTicks <= 0){
+			if($this->fireTicks <= 0 or $this->level->getWeather() == Level::WEATHER_RAIN or $this->level->getWeather() == Level::WEATHER_RAIN_THUNDER){
```
Prefer `===` over `==` as per our strict type checking preferences.

dktapps commented on src/pocketmine/Player.php:
```diff
@@ -1872,8 +1873,8 @@ protected function completeLoginSequence(){
-		$pk->rainLevel = 0; //TODO: implement these properly
-		$pk->lightningLevel = 0;
+		$pk->rainLevel = ($this->getLevel()->getWeather() === Level::WEATHER_RAIN or $this->getLevel()->getWeather() == Level::WEATHER_RAIN_THUNDER) ? 0 : 1;
+		$pk->lightningLevel = ($this->getLevel()->getWeather() === Level::WEATHER_RAIN_THUNDER) ? 0 : 1;
```
These aren't on/off, they are floats. Rain/thunder strength from 0-1.

dktapps commented on src/pocketmine/Player.php:
```diff
@@ -3590,6 +3591,25 @@ public function sendTranslation($message, array $parameters = []){
+		$pk2->data = mt_rand(30000, 40000);
+		$this->dataPacket($pk1);
+		$this->dataPacket($pk2);
+	}
```
This code is clearly copied from Genisys. I recognize this exact code.

jasonwynn10 replied:
I didn't intend to copy it explicitly. I was looking for a range of values and Genesis just happened to have some.

dktapps commented on src/pocketmine/Player.php:
```diff
@@ -3590,6 +3591,25 @@ public function sendTranslation($message, array $parameters = []){
+			$pk1->evid = LevelEventPacket::EVENT_STOP_RAIN;
+			$pk2->evid = LevelEventPacket::EVENT_STOP_THUNDER;
+		}
+		$pk1->data = mt_rand(90000, 110000); //If we're clearing the weather, it doesn't matter what strength values we set
```
These values are incorrect. There's some weird stuff with coercion, but these values are the rain/thunder strength (0f-1f) multiplied by 65535. The maximum values for these should be 65535.

jasonwynn10 replied:
How can the values multiplied by 65535 be lower or equal to 65535?

jasonwynn10 replied:
How can the values multiplied by 65535 be lower or equal to 65535? @dktapps

dktapps replied:
Simple maths. Float ranging from 0 to 1, multiplied by 65535.

jasonwynn10 replied:
Okay, so how does 0f and 1f fit in?

dktapps replied:
... standard representation for floats... it just means 0 - 1 as floats.

jasonwynn10 replied:
oh 😅 I'm clearly not thinking right now

dktapps reviewed (changes requested):
This pull request shares a similar fault with the other PR: you've hardcoded rain and thunder together. This reduces flexibility.

There are some unrelated changes in this pull request (submodule version changes), please remove those changes.

I've made some comments on the diff.

Please also make sure you take strict-typing into consideration and perform appropriate type checking and casts where necessary.

dktapps commented:
This also does not provide any flexibility whatsoever for varying rain and thunder strengths - you've hardcoded everything to an on-off switch.

For God's sake, it's using _mt_rand_. What happens if you have lots of players on the server and you send them weather like that? Everyone's going to see different weather strengths.

jasonwynn10 commented:
Oh yea... I didn't test with multiple players online 😅

jasonw4331 committed:
First Attempt to meet required specifications
Changes ToggleDownfall command class name
Added thunder separately
Corrected if statement type checking
Added weather intensity

jasonw4331 committed:
Changed return types and parameter types
Modified how the math is done to allow for easier plugin changes

jasonw4331 committed:
Missed these values

jasonw4331 committed:
These are better used

jasonw4331 committed:
Merge remote-tracking branch 'origin/master' into weather

jasonw4331 committed:
Fixed an issue where the data field wasn't an integer

jasonw4331 committed:
corrected some math

jasonwynn10 commented:
I broke something somewhere... 😅 
Unfortunately, I don't know when, where, or how. 🤷‍♂️

Qimie commented:
Just keep trying. You can do this!

On Jun 7, 2017 11:12 AM, "jasonwynn10" <notifications@github.com> wrote:

> I broke something somewhere... 😅
> Unfortunately, I don't know when, where, or how. 🤷‍♂️
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pmmp/PocketMine-MP/pull/460#issuecomment-306674207>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ARW33WKYrXQQ5Sbi-843QJNm-wwxi78fks5sBhUOgaJpZM4Mo82c>
> .
>

thebigsmileXD commented:
do you know the "??" operator?

also, as stated before, dont mtrand it. use the random class, combined with the tick of the level/server.

jasonwynn10 commented:
Obsolete
